lgc/patch/NggPrimShader.cpp

```diff
@@ -5998,6 +6010,70 @@ Value *NggPrimShader::doDppUpdate(Value *oldValue, Value *srcValue, unsigned dpp
                                      m_builder->getInt32(bankMask), m_builder->getInt1(boundCtrl)});
 }
 
+// =====================================================================================================================
+// Fetches the position data for the specified vertex ID.
+//
+// @param vertexId : Vertex thread ID in sub-group.
+Value *NggPrimShader::fetchVertexPositionData(Value *vertexId) {
+  if (m_hasGs) {
+    // ES-GS
+    auto &inOutUsage = m_pipelineState->getShaderResourceUsage(ShaderStageGeometry)->inOutUsage;
+    assert(inOutUsage.builtInOutputLocMap.find(BuiltInPosition) != inOutUsage.builtInOutputLocMap.end());
+    const unsigned loc = inOutUsage.builtInOutputLocMap[BuiltInPosition];
+    const unsigned rasterStream = inOutUsage.gs.rasterStream;
+
+    const unsigned vertexItemSize = 4 * inOutUsage.gs.outLocCount[rasterStream];
+    auto vertexOffset = m_builder->CreateMul(vertexId, m_builder->getInt32(vertexItemSize));
+    vertexOffset = m_builder->CreateShl(vertexOffset, 2);
+
+    return importGsOutput(VectorType::get(m_builder->getFloatTy(), 4), loc, 0, rasterStream, vertexOffset);
+  } else {
+    // ES-only
+    return readPerThreadDataFromLds(VectorType::get(m_builder->getFloatTy(), 4), vertexId, LdsRegionPosData);
```

Comment:
Can you flip the if condition and return early in case of `ES-only`?

Reply:
Yes, it is a good practice to use early returns for multiple simple checks. Here, I still make a if-else is just to make the logic clear: two branches, ES-GS and ES-only. And we will do a middle-scale change to modify ES-only NGG in the future. And the ES-only branch might not be just a simple return after that. More calculation might be involved :)

Reply:
Ack, thanks for the extra context.
In that case, I'd simply keep the same same if condition but remove the else, since the `if-then` must return anyway.

Reply:
To elaborate on what my point is: reaching the end of a non-void function without returning a value is an undefined behavior in C++, and these things are annoying to debug as they propagate internationally. When I look at a function and see a terminator at the end (`return` or `llvm_unreachable`), I don't have to worry at all about this particular kind of UB. When you write `if (...) { ... return X; } else { ... return Y; } }<end-of-function>` my worry is that even though the code looks correct right now, one of the return statements may be become conditional in the future, and the function could exit without returning.
I think this is why LLVM uses this style.

Reply:
Thanks for your valuable suggestions. I will change the code to make it always have return.